### ISSUE TITLE
Not able to use blob store from previous version with name containing space

### ISSUE THREAD

**Issue by goloxxly** [sonatype-nexus-community/nexus-blobstore-google-cloud#50]
* **What are you trying to do?**
I am trying to upgrade from version 0.4.0 to 0.9.1 keeping an existing Blob store called "Google what-ever"

I can see the following exceptions in the log:

```
2019-11-19 13:38:19,153+0000 ERROR [FelixStartLevel] *SYSTEM org.sonatype.nexus.repository.internal.blobstore.BlobStoreManagerImpl - Unable to restore BlobStore BlobStoreConfiguration{name='Google what-ever', type='Google Cloud Storage', attributes={google cloud storage={bucket=what-ever, credential_file=/secrets/google/storage-writer-access-key.json}}}
org.sonatype.nexus.blobstore.api.BlobStoreException: Unable to initialize blob store bucket: what-ever, Cause: namespace must the following pattern: [0-9A-Za-z\._\-]{0,100}
        at org.sonatype.nexus.blobstore.gcloud.internal.GoogleCloudBlobStore.doInit(GoogleCloudBlobStore.java:392)
        at org.sonatype.nexus.blobstore.BlobStoreSupport.init(BlobStoreSupport.java:219)
        at org.sonatype.nexus.repository.internal.blobstore.BlobStoreManagerImpl.doStart(BlobStoreManagerImpl.java:127)
        at org.sonatype.nexus.common.stateguard.StateGuardLifecycleSupport.start(StateGuardLifecycleSupport.java:67)
        at org.sonatype.nexus.repository.internal.blobstore.BlobStoreManagerImpl$$EnhancerByGuice$$a6075bb4.CGLIB$start$19(<generated>)
        at org.sonatype.nexus.repository.internal.blobstore.BlobStoreManagerImpl$$EnhancerByGuice$$a6075bb4$$FastClassByGuice$$313e7a33.invoke(<generated>)
        at com.google.inject.internal.cglib.proxy.$MethodProxy.invokeSuper(MethodProxy.java:228)
        at com.google.inject.internal.InterceptorStackCallback$InterceptedMethodInvocation.proceed(InterceptorStackCallback.java:76)
        at org.sonatype.nexus.common.stateguard.MethodInvocationAction.run(MethodInvocationAction.java:39)
        at org.sonatype.nexus.common.stateguard.StateGuard$TransitionImpl.run(StateGuard.java:193)
        at org.sonatype.nexus.common.stateguard.TransitionsInterceptor.invoke(TransitionsInterceptor.java:56)
        at com.google.inject.internal.InterceptorStackCallback$InterceptedMethodInvocation.proceed(InterceptorStackCallback.java:77)
        at com.google.inject.internal.InterceptorStackCallback.intercept(InterceptorStackCallback.java:55)
        at org.sonatype.nexus.repository.internal.blobstore.BlobStoreManagerImpl$$EnhancerByGuice$$a6075bb4.start(<generated>)
        at org.sonatype.nexus.repository.manager.internal.RepositoryManagerImpl.doStart(RepositoryManagerImpl.java:228)
        at org.sonatype.nexus.common.stateguard.StateGuardLifecycleSupport.start(StateGuardLifecycleSupport.java:67)
        at org.sonatype.nexus.repository.manager.internal.RepositoryManagerImpl$$EnhancerByGuice$$6b7b8b8d.CGLIB$start$21(<generated>)
        at org.sonatype.nexus.repository.manager.internal.RepositoryManagerImpl$$EnhancerByGuice$$6b7b8b8d$$FastClassByGuice$$aa1e4302.invoke(<generated>)
        at com.google.inject.internal.cglib.proxy.$MethodProxy.invokeSuper(MethodProxy.java:228)
        at com.google.inject.internal.InterceptorStackCallback$InterceptedMethodInvocation.proceed(InterceptorStackCallback.java:76)
        at org.sonatype.nexus.common.stateguard.MethodInvocationAction.run(MethodInvocationAction.java:39)
        at org.sonatype.nexus.common.stateguard.StateGuard$TransitionImpl.run(StateGuard.java:193)
        at org.sonatype.nexus.common.stateguard.TransitionsInterceptor.invoke(TransitionsInterceptor.java:56)
        at com.google.inject.internal.InterceptorStackCallback$InterceptedMethodInvocation.proceed(InterceptorStackCallback.java:77)
        at com.google.inject.internal.InterceptorStackCallback.intercept(InterceptorStackCallback.java:55)
        at org.sonatype.nexus.repository.manager.internal.RepositoryManagerImpl$$EnhancerByGuice$$6b7b8b8d.start(<generated>)
        at org.sonatype.nexus.extender.NexusLifecycleManager.startComponent(NexusLifecycleManager.java:199)
        at org.sonatype.nexus.extender.NexusLifecycleManager.to(NexusLifecycleManager.java:111)
        at org.sonatype.nexus.extender.NexusContextListener.moveToPhase(NexusContextListener.java:311)
        at org.sonatype.nexus.extender.NexusContextListener.frameworkEvent(NexusContextListener.java:208)
        at org.apache.felix.framework.Felix.setActiveStartLevel(Felix.java:1431)
        at org.apache.felix.framework.FrameworkStartLevelImpl.run(FrameworkStartLevelImpl.java:308)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalArgumentException: namespace must the following pattern: [0-9A-Za-z\._\-]{0,100}
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:135)
        at com.google.cloud.datastore.Validator.validateNamespace(Validator.java:54)
        at com.google.cloud.datastore.BaseKey$Builder.setNamespace(BaseKey.java:111)
        at org.sonatype.nexus.blobstore.gcloud.internal.DeletedBlobIndex.<init>(DeletedBlobIndex.java:69)
        at org.sonatype.nexus.blobstore.gcloud.internal.GoogleCloudBlobStore.doInit(GoogleCloudBlobStore.java:389)
        ... 32 common frames omitted
2019-11-19 13:38:19,939+0000 ERROR [FelixStartLevel] *SYSTEM org.sonatype.nexus.blobstore.gcloud.internal.ShardedCounterMetricsStore - Failed transition: NEW -> STARTED
java.lang.IllegalArgumentException: namespace must the following pattern: [0-9A-Za-z\._\-]{0,100}
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:135)
        at com.google.cloud.datastore.Validator.validateNamespace(Validator.java:54)
        at com.google.cloud.datastore.BaseKey$Builder.setNamespace(BaseKey.java:111)
        at org.sonatype.nexus.blobstore.gcloud.internal.ShardedCounterMetricsStore.doStart(ShardedCounterMetricsStore.java:156)
        at org.sonatype.nexus.common.stateguard.StateGuardLifecycleSupport.start(StateGuardLifecycleSupport.java:67)
        at org.sonatype.nexus.common.stateguard.MethodInvocationAction.run(MethodInvocationAction.java:39)
        at org.sonatype.nexus.common.stateguard.StateGuard$TransitionImpl.run(StateGuard.java:193)
        at org.sonatype.nexus.common.stateguard.TransitionsInterceptor.invoke(TransitionsInterceptor.java:56)
        at org.sonatype.nexus.blobstore.gcloud.internal.GoogleCloudBlobStore.doStart(GoogleCloudBlobStore.java:183)
        at org.sonatype.nexus.common.stateguard.StateGuardLifecycleSupport.start(StateGuardLifecycleSupport.java:67)
        at org.sonatype.nexus.common.stateguard.MethodInvocationAction.run(MethodInvocationAction.java:39)
        at org.sonatype.nexus.common.stateguard.StateGuard$TransitionImpl.run(StateGuard.java:193)
        at org.sonatype.nexus.common.stateguard.TransitionsInterceptor.invoke(TransitionsInterceptor.java:56)
        at org.sonatype.nexus.repository.internal.blobstore.BlobStoreManagerImpl.doStart(BlobStoreManagerImpl.java:145)
        at org.sonatype.nexus.common.stateguard.StateGuardLifecycleSupport.start(StateGuardLifecycleSupport.java:67)
        at org.sonatype.nexus.common.stateguard.MethodInvocationAction.run(MethodInvocationAction.java:39)
        at org.sonatype.nexus.common.stateguard.StateGuard$TransitionImpl.run(StateGuard.java:193)
        at org.sonatype.nexus.common.stateguard.TransitionsInterceptor.invoke(TransitionsInterceptor.java:56)
        at org.sonatype.nexus.repository.manager.internal.RepositoryManagerImpl.doStart(RepositoryManagerImpl.java:228)
        at org.sonatype.nexus.common.stateguard.StateGuardLifecycleSupport.start(StateGuardLifecycleSupport.java:67)
        at org.sonatype.nexus.common.stateguard.MethodInvocationAction.run(MethodInvocationAction.java:39)
        at org.sonatype.nexus.common.stateguard.StateGuard$TransitionImpl.run(StateGuard.java:193)
        at org.sonatype.nexus.common.stateguard.TransitionsInterceptor.invoke(TransitionsInterceptor.java:56)
        at org.sonatype.nexus.extender.NexusLifecycleManager.startComponent(NexusLifecycleManager.java:199)
        at org.sonatype.nexus.extender.NexusLifecycleManager.to(NexusLifecycleManager.java:111)
        at org.sonatype.nexus.extender.NexusContextListener.moveToPhase(NexusContextListener.java:311)
        at org.sonatype.nexus.extender.NexusContextListener.frameworkEvent(NexusContextListener.java:208)
        at org.apache.felix.framework.Felix.setActiveStartLevel(Felix.java:1431)
        at org.apache.felix.framework.FrameworkStartLevelImpl.run(FrameworkStartLevelImpl.java:308)
        at java.lang.Thread.run(Thread.java:748)
2019-11-19 13:38:19,946+0000 ERROR [FelixStartLevel] *SYSTEM org.sonatype.nexus.blobstore.gcloud.internal.GoogleCloudBlobStore - Failed transition: NEW -> STARTED
java.lang.IllegalArgumentException: namespace must the following pattern: [0-9A-Za-z\._\-]{0,100}
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:135)
        at com.google.cloud.datastore.Validator.validateNamespace(Validator.java:54)
        at com.google.cloud.datastore.BaseKey$Builder.setNamespace(BaseKey.java:111)
        at org.sonatype.nexus.blobstore.gcloud.internal.ShardedCounterMetricsStore.doStart(ShardedCounterMetricsStore.java:156)
        at org.sonatype.nexus.common.stateguard.StateGuardLifecycleSupport.start(StateGuardLifecycleSupport.java:67)
        at org.sonatype.nexus.common.stateguard.MethodInvocationAction.run(MethodInvocationAction.java:39)
        at org.sonatype.nexus.common.stateguard.StateGuard$TransitionImpl.run(StateGuard.java:193)
        at org.sonatype.nexus.common.stateguard.TransitionsInterceptor.invoke(TransitionsInterceptor.java:56)
        at org.sonatype.nexus.blobstore.gcloud.internal.GoogleCloudBlobStore.doStart(GoogleCloudBlobStore.java:183)
        at org.sonatype.nexus.common.stateguard.StateGuardLifecycleSupport.start(StateGuardLifecycleSupport.java:67)
        at org.sonatype.nexus.common.stateguard.MethodInvocationAction.run(MethodInvocationAction.java:39)
        at org.sonatype.nexus.common.stateguard.StateGuard$TransitionImpl.run(StateGuard.java:193)
        at org.sonatype.nexus.common.stateguard.TransitionsInterceptor.invoke(TransitionsInterceptor.java:56)
        at org.sonatype.nexus.repository.internal.blobstore.BlobStoreManagerImpl.doStart(BlobStoreManagerImpl.java:145)
        at org.sonatype.nexus.common.stateguard.StateGuardLifecycleSupport.start(StateGuardLifecycleSupport.java:67)
        at org.sonatype.nexus.common.stateguard.MethodInvocationAction.run(MethodInvocationAction.java:39)
        at org.sonatype.nexus.common.stateguard.StateGuard$TransitionImpl.run(StateGuard.java:193)
        at org.sonatype.nexus.common.stateguard.TransitionsInterceptor.invoke(TransitionsInterceptor.java:56)
        at org.sonatype.nexus.repository.manager.internal.RepositoryManagerImpl.doStart(RepositoryManagerImpl.java:228)
        at org.sonatype.nexus.common.stateguard.StateGuardLifecycleSupport.start(StateGuardLifecycleSupport.java:67)
        at org.sonatype.nexus.common.stateguard.MethodInvocationAction.run(MethodInvocationAction.java:39)
        at org.sonatype.nexus.common.stateguard.StateGuard$TransitionImpl.run(StateGuard.java:193)
        at org.sonatype.nexus.common.stateguard.TransitionsInterceptor.invoke(TransitionsInterceptor.java:56)
        at org.sonatype.nexus.extender.NexusLifecycleManager.startComponent(NexusLifecycleManager.java:199)
        at org.sonatype.nexus.extender.NexusLifecycleManager.to(NexusLifecycleManager.java:111)
        at org.sonatype.nexus.extender.NexusContextListener.moveToPhase(NexusContextListener.java:311)
        at org.sonatype.nexus.extender.NexusContextListener.frameworkEvent(NexusContextListener.java:208)
        at org.apache.felix.framework.Felix.setActiveStartLevel(Felix.java:1431)
        at org.apache.felix.framework.FrameworkStartLevelImpl.run(FrameworkStartLevelImpl.java:308)
        at java.lang.Thread.run(Thread.java:748)

```

* **What feature or behavior is this required for?**

* **How could we solve this issue? (Not knowing is okay!)**
Maybe
```
namespace=blobstore-[BlobStoreConfiguration.getName()]
```
should replace unsupported chars with something that's allowed. 

* **Anything else?**
Any quick work-around would be appreciated.

**Comment by nblair** [comment]
Interesting, thanks for the report @goloxxly. Sadly I don't have a quick work-around. Given the experimental nature of this plugin, I haven't really developed any migration for the blobstore metadata that moved from Storage to Datastore in 0.8.0.

I haven't used a bucket with the `-` in the name, but now I know that's something we will need to validate for the future (since Storage allows it for bucket names but Datastore doesn't allow it for namespace).

How large is the blobstore you are looking to upgrade?

**Comment by goloxxly** [comment]
As the error says:
```
namespace must the following pattern: [0-9A-Za-z\._\-]{0,100}
```
I believe the problem is not the hyphen, but the space.

The blobstore is around 10TB.

**Comment by nblair** [comment]
Let me see what I can do; with a night's sleep I have some ideas.

**Comment by nblair** [comment]
If a path forward required you to rename the blobstore configuration (not the bucket in GCP) without the space, would that be acceptable?

**Comment by goloxxly** [comment]
I'm sorry but I don't think that it would be possible to rename a blobstore in Nexus.

**Comment by nblair** [comment]
This was closed automatically from the reference to #52 being merged. 
I hope the workaround on creating a new blobstore is ok, please re-open here if it isn't.
I'll focus on #51 if that's a bigger priority.
